Add PocketIC testing

ulan opened the pull request:
This adds a new `tests` directory with PocketIC integration tests.

Summary of changes:

- Add a script to download `pocket-ic` binary (tested only on linux).
- Add a `make pocket_ic` rule.
- Adjust the `make test` rule to first build the Wasm binaries and call `cargo test`.
- Add tests for the backend and bucket canisters.

ulan committed:
Add PocketIC testing

This adds a new `tests` directory with PocketIC integration tests.

Summary of changes:

- Add a script to download `pocket-ic` binary (tested only on linux).
- Add a `make pocket_ic` rule.
- Adjust the `make test` rule to first build the Wasm binaries and call `cargo test`.
- Add tests for the backend and bucket canisters.

ulan committed:
Install pocket-ic in github workflow

taggrx commented on tests/src/backend_tests.rs:
```diff
@@ -0,0 +1,131 @@
+}
+
+#[test]
+fn test_upgrades() {
```
I would definitely keep this test as it serves as a great example for me when I'll add more of these tests, but I wonder if you knew that our e2e tests do the same but via UI and even doing a second upgrade using the recovery mechanism?

xqxpx reviewed (approved):
Thanks so much Ulan, this is was indeed a missing piece so far: there is functionality which cannot be really tested via e2e tests like distributions and cannot be tested as a unit test. 

Do you agree that we should use these tests especially for this kind of functionalities?

ulan committed:
- Adjust the backup script to work on Linux
- Adjust tests to skip running if the Wasm binary doesn't have
  dev endpoints.
- Add a test for upgrade based on backup data that runs only
  if the data is present in `tests/snapshot` directory.

ulan committed:
Merge branch 'TaggrNetwork:main' into pocket-ic

ulan committed:
Fix formatting
